src/v/raft/consensus.cc

```diff
@@ -557,41 +565,55 @@ ss::future<result<replicate_result>> consensus::do_replicate(
+                  return seastar::make_ready_future<result<replicate_result>>(
+                    errc::not_leader);
+              }
 
-          if (expected_term.has_value() && expected_term.value() != _term) {
-              return seastar::make_ready_future<result<replicate_result>>(
-                errc::not_leader);
-          }
-          _last_write_consistency_level = lvl;
-          return disk_append(
-                   model::make_record_batch_reader<
-                     details::term_assigning_reader>(
-                     std::move(rdr), model::term_id(_term)),
-                   update_last_quorum_index::no)
-            .then([this](storage::append_result res) {
-                // only update visibility upper bound if all quorum replicated
-                // entries are committed already
-                if (_commit_index >= _last_quorum_replicated_index) {
-                    // for relaxed consistency mode update visibility upper
-                    // bound with last offset appended to the log
-                    _visibility_upper_bound_index = std::max(
-                      _visibility_upper_bound_index, res.last_offset);
-                    maybe_update_majority_replicated_index();
-                }
-                return result<replicate_result>(
-                  replicate_result{.last_offset = res.last_offset});
-            });
-      })
-      .finally([this] { _probe.replicate_done(); });
+              if (expected_term.has_value() && expected_term.value() != _term) {
+                  return seastar::make_ready_future<result<replicate_result>>(
+                    errc::not_leader);
+              }
+              _last_write_consistency_level = lvl;
+              return disk_append(
+                       model::make_record_batch_reader<
+                         details::term_assigning_reader>(
+                         std::move(rdr), model::term_id(_term)),
+                       update_last_quorum_index::no)
+                .then([this](storage::append_result res) {
+                    // only update visibility upper bound if all quorum
+                    // replicated entries are committed already
+                    if (_commit_index >= _last_quorum_replicated_index) {
+                        // for relaxed consistency mode update visibility upper
+                        // bound with last offset appended to the log
+                        _visibility_upper_bound_index = std::max(
+                          _visibility_upper_bound_index, res.last_offset);
+                        maybe_update_majority_replicated_index();
+                    }
+                    return result<replicate_result>(
+                      replicate_result{.last_offset = res.last_offset});
+                })
+                .finally([this, u = std::move(u)] { _probe.replicate_done(); });
+          });
+      });
+
+    return replicate_stages(std::move(enqueued_f), std::move(replicated));
```

Comment:
does `enqueued_f` also need to be wrapped in the gate?

Reply:
I've change the way we handle gate in current version.